python/sgl_jax/srt/models/qwen2_moe.py

```diff
@@ -0,0 +1,605 @@
+                    weight_dtype=dtype,
+                    dtype=dtype,
+                    mesh=mesh,
+                )
+
+        self.input_layernorm = RMSNorm(
+            config.hidden_size,
+            epsilon=config.rms_norm_eps,
+            param_dtype=dtype,
+            scale_init=nnx.with_partitioning(init_fn, (None,)),
+            rngs=rngs,
+        )
+        self.post_attention_layernorm = RMSNorm(
+            config.hidden_size,
+            epsilon=config.rms_norm_eps,
+            param_dtype=dtype,
+            scale_init=nnx.with_partitioning(init_fn, (None,)),
+            rngs=rngs,
+        )
+
+    def __call__(
+        self,
+        positions: jax.Array,
+        hidden_states: jax.Array,
+        forward_batch: ForwardBatch,
+        token_to_kv_pool: KVCache,
+        residual: jax.Array | None = None,
+    ) -> tuple[jax.Array, jax.Array]:
+        layer_callback_flag = []
+        if residual is None:
+            residual = hidden_states
+            hidden_states = self.input_layernorm(hidden_states)
+        else:
+            hidden_states += residual
+            residual = hidden_states
+            hidden_states = self.input_layernorm(hidden_states)
+
+        hidden_states, kv_fused = self.self_attn(
+            positions=positions,
+            hidden_states=hidden_states,
+            forward_batch=forward_batch,
+            token_to_kv_pool=token_to_kv_pool,
+        )
+
+        hidden_states += residual
+        residual = hidden_states
+        hidden_states = self.post_attention_layernorm(hidden_states)
+
+        # optional shared expert output
+        if self.shared_experts is not None:
+            shared_output = self.shared_experts(hidden_states)
+            if self.shared_expert_gate is not None:
+                gate, _ = self.shared_expert_gate(hidden_states)
+                shared_output = jax.nn.sigmoid(gate) * shared_output
+        else:
+            shared_output = None
+
+        router_logits = self.moe_gate(hidden_states)
+        topk_weights, topk_ids = self.topk(router_logits)
+        mlp_output = self.mlp(hidden_states, topk_weights, topk_ids)
+        hidden_states = mlp_output if shared_output is None else (mlp_output + shared_output)
+
+        return hidden_states, residual, kv_fused, layer_callback_flag
+
+
+class Qwen2MoeModel(nnx.Module):
+    """Qwen2 MoE model with embedding, layers, and normalization."""
+
+    def __init__(
+        self,
+        config: PretrainedConfig,
+        dtype: jnp.dtype = jnp.bfloat16,
+        rngs: nnx.Rngs = None,
+        mesh: jax.sharding.Mesh = None,
+    ):
+        self.config = config
+        self.padding_idx = config.pad_token_id
+        self.vocab_size = config.vocab_size
+
+        self.embed_tokens = Embed(
+            num_embeddings=config.vocab_size,
+            features=config.hidden_size,
+            rngs=rngs,
+            dtype=dtype,
+            param_dtype=dtype,
+        )
```

Comment:
The current default configuration performs TP splitting on the vocab size dimension for both embed tokens and LM heads. Please verify and modify as needed.